src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCISharedBuildJobQueueService.java

```diff
@@ -337,33 +342,51 @@ private void processBuild(LocalCIBuildJobQueueItem buildJob) {
             checkAvailabilityAndProcessNextBuild();
         }).exceptionally(ex -> {
-            log.error("Error while processing build job: {}", buildJob, ex);
+            if (ex.getCause() instanceof CancellationException && ex.getMessage().equals("Build job with id " + buildJob.getId() + " was cancelled.")) {
+                localProcessingJobs.decrementAndGet();
+                updateLocalBuildAgentInformation();
+            }
+            else {
+                log.error("Error while processing build job: {}", buildJob, ex);
 
-            processingJobs.remove(buildJob.getId());
-            localProcessingJobs.decrementAndGet();
-            updateLocalBuildAgentInformation();
+                processingJobs.remove(buildJob.getId());
+                localProcessingJobs.decrementAndGet();
+                updateLocalBuildAgentInformation();
 
-            if (buildJob.getRetryCount() > 0) {
-                log.error("Build job failed for the second time: {}", buildJob);
-                return null;
-            }
+                if (buildJob.getRetryCount() > 0) {
+                    log.error("Build job failed for the second time: {}", buildJob);
+                    return null;
+                }
 
-            // Do not requeue the build job if the participation has been deleted in the meantime
-            SecurityUtils.setAuthorizationObject();
-            Optional<Participation> participationOptional = participationRepository.findById(participation.getId());
-            if (participationOptional.isPresent()) {
-                log.warn("Requeueing failed build job: {}", buildJob);
-                buildJob.setRetryCount(buildJob.getRetryCount() + 1);
-                queue.add(buildJob);
-                checkAvailabilityAndProcessNextBuild();
-            }
-            else {
-                log.warn("Participation with id {} has been deleted. Cancelling the requeueing of the build job.", participation.getId());
+                // Do not requeue the build job if the participation has been deleted in the meantime
+                SecurityUtils.setAuthorizationObject();
+                Optional<Participation> participationOptional = participationRepository.findById(participation.getId());
+                if (participationOptional.isPresent()) {
+                    log.warn("Requeueing failed build job: {}", buildJob);
+                    buildJob.setRetryCount(buildJob.getRetryCount() + 1);
+                    queue.add(buildJob);
+                    checkAvailabilityAndProcessNextBuild();
+                }
+                else {
+                    log.warn("Participation with id {} has been deleted. Cancelling the requeueing of the build job.", participation.getId());
+                }
             }
             return null;
         });
     }
 
+    private void notifyUserAboutSubmissionError(ProgrammingExerciseParticipation participation, BuildTriggerWebsocketError error) {
```

Comment:
Maybe create a method for this in ProgrammingMessagingService to avoid code duplication